Codec/Archive/Tar/Types.hs

```diff
@@ -462,46 +467,48 @@ newtype LinkTarget = LinkTarget BS.ByteString
 instance NFData LinkTarget where
     rnf (LinkTarget bs) = rnf bs
 
--- | Convert a native 'FilePath' to a tar 'LinkTarget'. This may fail if the
+-- | Convert a native 'FilePath' to a tar 'LinkTarget'.
 -- string is longer than 100 characters or if it contains non-portable
 -- characters.
---
-toLinkTarget   :: FilePath -> Maybe LinkTarget
-toLinkTarget path | length path <= 100 = Just $! LinkTarget (packAscii path)
-                  | otherwise          = Nothing
+toLinkTarget   :: MonadThrow m => FilePath -> m LinkTarget
+toLinkTarget path | length path <= 100 = do
+                                           target <- toLinkTarget' path
+                                           pure $! LinkTarget (packAscii target)
+                  | otherwise          = throwM (TooLong path)
+
+data LinkTargetException = IsAbsolute FilePath
+                         | TooLong FilePath
+  deriving (Show,Typeable)
+
+instance Exception LinkTargetException where
+  displayException (IsAbsolute fp) = "Link target \"" <> fp <> "\" is unexpectedly absolute"
+  displayException (TooLong _) = "The link target is too long"
+
+-- | Convert a native 'FilePath' to a unix filepath suitable for
+-- using as 'LinkTarget'. Does not error if longer than 100 characters.
+toLinkTarget' :: MonadThrow m => FilePath -> m FilePath
+toLinkTarget' path
+  | FilePath.Native.isAbsolute path = throwM (IsAbsolute path)
```

Comment:
You could argue this is duplication with `checkEntrySecurity`, but we really can't convert safely if it's an absolute path. I prefer to error early.